### PR TITLE
Update `NeptuneLogger`

### PR DESCRIPTION
**Pull request by AleksanderWWW**
# What does this PR do?

- [x] Update the parameter name `upload_artifacts` -> `upload_checkpoints`
- [x] Improve the way the `OOMCallback` handles files
- [x] Better handle normalized images logging
- [x] Improve file uploading
- [x] Improve docs
- [x] Add optional progress bar support for file downloading (if Neptune version permits)

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by AleksanderWWW** [comment]
https://github.com/mosaicml/composer/issues/3170

@SiddhantSadangi @normandy7

**Review by mvpatel2000 — COMMENTED** [review]
LGTM, though would like Cheng to review too given other changes

**Comment by mvpatel2000** [comment]
@AleksanderWWW  can you please provide a proper PR description?

**Review by cli99 — COMMENTED** [review]
LGTM.

**Comment by mvpatel2000** [comment]
@AleksanderWWW  can approve and merge once tests pass. Let me know if you have any trouble

**Comment by AleksanderWWW** [comment]
Idk, tried so many things but no idea how to fix this doctest failure due to unexpected output generated by neptune @mvpatel2000 
@SiddhantSadangi

File: docs/source/trainer/file_uploading.rst [hunk 1]
```diff
@@ -161,6 +161,33 @@ Weights & Biases Artifacts
...
+
+.. testcode::
+    :skipif: not _NEPTUNE_INSTALLED
+    import logging
+
+    from composer.loggers import NeptuneLogger
+    from composer import Trainer
+
+    logging.getLogger("neptune").setLevel(logging.CRITICAL)
+
```
Comment (SiddhantSadangi) [inline docs/source/trainer/file_uploading.rst:179]: We shouldn't be suppressing outputs in the docs code

```suggestion
    from composer.loggers import NeptuneLogger
    from composer import Trainer

```

File: docs/source/trainer/logging.rst [hunk 2]
```diff
@@ -29,34 +29,41 @@ and also saves them to the file
...
 
 .. testcode::
-    :skipif: not _WANDB_INSTALLED or not _COMETML_INSTALLED
-
+    :skipif: not _WANDB_INSTALLED or not _COMETML_INSTALLED or not _NEPTUNE_INSTALLED
+    import logging
     from composer import Trainer
     from composer.loggers import WandBLogger, CometMLLogger, MLFlowLogger, NeptuneLogger, FileLogger
 
+    logging.getLogger("neptune").setLevel(logging.CRITICAL)
```
Comment (SiddhantSadangi) [inline docs/source/trainer/logging.rst:50]: Same as above

```suggestion
```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]
Let's remove suppressing logs before we forget doing so :)

**Comment by SiddhantSadangi** [comment]
The error doesn't seem to related to Neptune outputs
`ERROR: Error in "testcode" directive: invalid option block`

The block is the same as our previous PR that cleared tests though 🤷

File: docs/source/trainer/logging.rst (hunk 2 above)
Comment (SiddhantSadangi) [inline docs/source/trainer/logging.rst:50]: Same as before
```suggestion
    from composer import Trainer
    from composer.loggers import WandBLogger, CometMLLogger, MLFlowLogger, NeptuneLogger, FileLogger

```

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

**Comment by AleksanderWWW** [comment]
Ok, idk how, but removing `logging.getLogger` and level setting + adding a whitespace here and there seemed to resolve the issue. I will definetely wake up tomorrow morning a little wiser and a lot more confused than I used to be.

File: composer/callbacks/oom_observer.py [hunk 3]
```diff
@@ -136,31 +159,26 @@ def oom_observer(device: int, alloc: int, device_alloc: int, device_free: int):
...
 
                 if self.remote_path_in_bucket is not None:
-                    for f in [
-                        snapshot_file,
-                        trace_plot_file,
-                        segment_plot_file,
-                        segment_flamegraph_file,
-                        memory_flamegraph_file,
-                    ]:
+
```
Comment (mvpatel2000) [inline composer/callbacks/oom_observer.py:180]:
```suggestion
```

File: tests/loggers/test_neptune_logger.py [hunk 4]
```diff
@@ -153,3 +155,55 @@ def test_neptune_log_image(test_neptune_logger):
...
+        np.ones((4, 4)) * -1,
+        np.identity(4) * 300 - 1,
+    ]
+
+    for image in image_variants:
+        test_neptune_logger.log_images(images=image)
+
+        mock_scale_img.assert_called_once()
+
+        mock_scale_img.reset_mock()
```
Comment (mvpatel2000) [inline tests/loggers/test_neptune_logger.py:209]:
```suggestion
    for image in image_variants:
        test_neptune_logger.log_images(images=image)
        mock_scale_img.assert_called_once()
        mock_scale_img.reset_mock()
```

File: tests/loggers/test_neptune_logger.py [hunk 5]
```diff
@@ -153,3 +155,55 @@ def test_neptune_log_image(test_neptune_logger):
...
+
+    os.remove(symlink_name)
+    os.remove(file_name)
+
+
+def test_neptune_log_image_warns_about_improper_value_range(test_neptune_logger):
+    image = np.ones((4, 4)) * 300
+    with pytest.warns() as record:
+        test_neptune_logger.log_images(images=image)
+
```
Comment (mvpatel2000) [inline tests/loggers/test_neptune_logger.py:192]:
```suggestion
```

File: tests/loggers/test_neptune_logger.py [hunk 6]
```diff
@@ -153,3 +155,55 @@ def test_neptune_log_image(test_neptune_logger):
...
+def test_neptune_logger_doesnt_upload_symlinks(test_neptune_logger, dummy_state):
+    with _manage_symlink_creation('test.txt') as symlink_name:
+        test_neptune_logger.upload_file(
+            state=dummy_state,
+            remote_file_name='test_symlink',
+            file_path=Path(symlink_name),
+        )
+
+    assert not test_neptune_logger.neptune_run.exists(f'{test_neptune_logger._base_namespace}/test_symlink')
```
Comment (mvpatel2000) [inline tests/loggers/test_neptune_logger.py:168]:
```suggestion
def test_neptune_logger_doesnt_upload_symlinks(test_neptune_logger, dummy_state):
    with _manage_symlink_creation('test.txt') as symlink_name:
        test_neptune_logger.upload_file(
            state=dummy_state,
            remote_file_name='test_symlink',
            file_path=Path(symlink_name),
        )
    assert not test_neptune_logger.neptune_run.exists(f'{test_neptune_logger._base_namespace}/test_symlink')
```

**Review by mvpatel2000 — APPROVED** [review]
LGTM, a few minor nits I will apply

**Comment by mvpatel2000** [comment]
@AleksanderWWW  the following issue is causing test failures:
```
Expected nothing
Got:
Warning:     [neptune] [warning] Info (NVML): NVML Shared Library Not Found. GPU usage metrics may not be reported. For more information, see https://docs.neptune.ai/help/nvml_error/
```
is there a way to pass args to neptune to not emit warnings in this case? or does it have to be filtered

**Comment by mvpatel2000** [comment]
Hm... it might be flakey... I reran and it's gone 🤔. I will rerun one more time. If it passes we can merge and if it flakes post-merge ill revert and reopen PR

**Pull request merged**